File: wordsearch.go
// Time Complexity : O(mn3Pow(wordlen))
// Space Complexity : O(len(word))
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
create dirs array for 4 directions
for each row and column call recursive with i, j, index 0
if any one is true return true

in recursive function
if index == len(word) return true
else if position out of range or value on board == # return false

if word[index] == board[i][j]
get value of board at position in a temp
for each direction call recursive again with next index and if any one is true return true
replace board at position
if none satisfy return false
*/
package main

import "fmt"

var m,n int
var dirs [][]int
func exist(board [][]byte, word string) bool {
	m = len(board)
	n = len(board[0])
	dirs = [][]int{{0,1}, {-1,0}, {0,-1}, {1,0}}
	for i:=0;i<m;i++ {
		for j:=0;j<n;j++ {
			if h1(board, word, i, j, 0) {
				return true
			}
		}
	}
	return false
}

func h1(board [][]byte, word string, i, j, index int) bool {
	if index == len(word) {
		return true
	}
	if i < 0 || j < 0 || i >= m || j >= n || board[i][j] == '#' {
		return false
	}

	if word[index] == board[i][j] {
		temp := board[i][j]
		board[i][j] = '#'
		for x:=0;x<len(dirs);x++ {
			r := i + dirs[x][0]
			c := j + dirs[x][1]
			if h1(board, word, r, c, index + 1) {
				return true
			}
		}
		board[i][j] = temp
	}
	return false
}


func MainWordSearch() {
	fmt.Println(exist([][]byte{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}}, "ABCCED")) //expected true
}

File: WordSearch.cpp
//Time Complexity: 
//Space Complexity: 

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        char firstLetter = word[0];
        vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size(), false));
        for(int i = 0; i < board.size(); i++) {
            for(int j = 0; j < board[0].size(); j++) {
                if(board[i][j] == firstLetter) {
                    if(wordExists(i, j, board, visited, word, 0)) return true;  
                }
            }
        }
        return false;     
    }

    bool wordExists(int row, int col, vector<vector<char>>& board, vector<vector<bool>>& visited, string word, int wordIndex) {
        if(wordIndex >= (int)word.size()) return true; 
        
        if(row < 0 || row >= board.size() || col < 0 || col >= board[0].size()) return false; 

        if(board[row][col] == word[wordIndex] && !visited[row][col]) {
            visited[row][col] = true; 
            if(wordExists(row+1, col, board, visited, word, wordIndex+1)) return true; 
            if(wordExists(row-1, col, board, visited, word, wordIndex+1)) return true; 
            if(wordExists(row, col+1, board, visited, word, wordIndex+1)) return true;
            if(wordExists(row, col-1, board, visited, word, wordIndex+1)) return true;   
            visited[row][col] = false;                       
        }
        return false; 
    }
};